src/error.rs

```diff
@@ -0,0 +1,186 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at https://mozilla.org/MPL/2.0/.
+
+use std::{fmt, result};
+
+use failure::{Backtrace, Context, Fail};
+use rocket::{
+    self,
+    http::Status,
+    response::{self, Responder, Response},
+    Request,
+};
+use rocket_contrib::Json;
+
+use auth_db::BounceRecord;
+use logging::MozlogLogger;
+
+pub type AppResult<T> = result::Result<T, AppError>;
+
+#[derive(Debug)]
+pub struct AppError {
+    inner: Context<AppErrorKind>,
+}
+
+impl AppError {
+    pub fn kind(&self) -> &AppErrorKind {
+        self.inner.get_context()
+    }
+}
+
+impl Fail for AppError {
+    fn cause(&self) -> Option<&Fail> {
+        self.inner.cause()
+    }
+
+    fn backtrace(&self) -> Option<&Backtrace> {
+        self.inner.backtrace()
+    }
+}
+
+impl fmt::Display for AppError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.inner.fmt(f)
+    }
+}
+
+/// The specific kind of error that can occur.
+#[derive(AsRefStr, Clone, Debug, Eq, PartialEq)]
+pub enum AppErrorKind {
+    /// 400 Bad Request
+    BadRequest,
+    /// 404 Not Found
+    NotFound,
+    /// 405 Method Not Allowed
+    MethodNotAllowed,
+    /// 422 Unprocessable Entity
+    UnprocessableEntity,
+    /// 429 Too Many Requests,
+    TooManyRequests,
+    /// 500 Internal Server Error
+    InternalServerError,
```

Comment:
Seeing both these and our internal error types modelled inside `AppErrorKind` gives me a confusing mental model. Do we need to model the HTTP error types here, or do they all just fall out naturally if we do the appropriate work when serializing/mapping our internal types?

(not sure if that's clear, I mean if the implementations of `fmt` and `http_status` are correct for all our internal errors and we only return those internally, would they ever get invoked with these values?)

Reply:
Hmm, I understand what you're saying. I didn't really think of that, but it makes sense.

I'm not sure about what is the answer for this right now, but I will investigate and let you know.

Reply:
I tried removing these handlers, but what happens is that it falls back to the default rocket handler for them and we get a HTML response and no logging. I'll investigate this a little further, but I think we have to have these, even though they seem redundant.